[MRG+1] Novelty detection for LOF

albertcthomas opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

Fixes #10191

#### What does this implement/fix? Explain your changes.
This PR will allow users to use LOF for novelty detection as suggested in [this comment](https://github.com/scikit-learn/scikit-learn/pull/9015#issuecomment-347445677) of PR #9015.  An argument `novelty=True` or `False` is added in `__init__`:
- if `novelty=False` (default) then `lof.predict(X)` raises `ValueError('predict is not available when novelty=False, use fit_predict if you want to predict on training data. Use novelty=True if you want to use LOF for novelty detection and predict on new unseen data.')`
Similar error for `lof.decision_function(X)` and `lof.score_samples(X)`
- if `novelty=True`, `lof.fit_predict(X)` raises `ValueError('fit_predict is not available when novelty=True. Use novelty=False if you want to predict on the training set.')`. But `lof.predict(X)`, `lof.decision_function` and `lof.score_samples(X)` are OK.

This way users can use LOF for novelty detection but by default, as `novelty=False`, users will get an error with an informative message saying to use `novelty=True` if they really want to predict on new data.

#### Any other comments?
This PR also refactors doc on outlier detection: particularly, plot_outlier_detection.py is removed in favor of plot_anomaly_comparison.py


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

albertcthomas committed:
start novelty detection for lof

jnothman reviewed (commented):
So we have two big questions:
* does LOF make sense as an inductive model?
* if so, is it necessary to make two different modes?

It doesn't look like the fitting is any different in the inductive case from `fit_predict`. Is the main concern that prediction on the training data is not identical to the results of `fit_predict`? We have such cases with `{fit_,}transform` in the works, FWIW.

albertcthomas commented:
The main concern is indeed that `fit_predict` would not be identical to `fit.predict`.

jnothman commented:
I don't mind having such cases and just documenting them clearly. Maybe we
need a name for such a thing so that we can stamp such estimators clearly
with that name. Maybe we also need a warning when it looks like the user is
passing the training data to predict... (By storing the training id(X) on
the estimator, but it might be hard to check if only a subset is being
passed in, be it a slice or certainly a copy)

agramfort commented on sklearn/neighbors/lof.py:
```diff
@@ -308,20 +356,29 @@ def _score_samples(self, X):
+               'considering the negative_outlier_factor_ attribute. Use '
+               'novelty=True if you want to use LOF for novelty detection and '
+               'compute score_samples for new unseen data.')
+        raise NotImplementedError(msg)
```
I would structure the code differently.

```
if not self.novelty:
     raise

...

return ...```

it feels better to first capture errors and to end the function with a return

agramfort commented:
I find this solution very elegant and rigorous. Using the id will not be a robust solution.
And we still don't break the contract that fit.predict == fit_predict unless the user explicitly asks for it.

jnothman commented on sklearn/neighbors/lof.py:
```diff
@@ -169,6 +177,14 @@ def fit_predict(self, X, y=None):
+        if self.novelty:
+            msg = ('fit_predict is not available when novelty=True. Use '
+                   'novelty=False if you want to predict on the training set.')
+            raise NotImplementedError(msg)
```
I don't think this is the right error type. It makes it seem like one day we will implement this case

jnothman replied:
A TypeError would be more normal for a bad call, while a ValueError would be possible too. None is quite satisfying.

agramfort replied:
I would use ValueError

jnothman commented:
Test failures. I'm happy with this solution. I'm still not sure we can maintain similar assurances for all Transformers, though

agramfort commented on sklearn/neighbors/lof.py:
```diff
@@ -250,26 +262,52 @@ def _predict(self, X=None):
+               'if you want to predict on training data. Use novelty=True if '
+               'you want to use LOF for novelty detection and predict on new '
+               'unseen data.')
+        raise NotImplementedError(msg)
```
ValueError

as I said elsewhere I think it more natural to finish a function with a return than a raise though

albertcthomas commented:
Thanks for your opinion and reviews @jnothman and @agramfort. I think this is a good solution as some people are not ready to break fit.predict = fit_predict (for what I think are good reasons). I'm OK with ValueError. In a way the error is related to the value of the `novelty` parameter in this case. I will finish this PR, I just wanted to be sure about this before completely implementing this feature.

jnothman commented:
are we sure about the name of the parameter?

agramfort commented:
that's fine with me.

albertcthomas commented:
Unless this kind of solution can be more generally applied (i.e. not just for LOF of outlier detection estimators), novelty seems good to me as well.

albertcthomas committed:
use ValueError and always raise exception before return in methods

albertcthomas committed:
fix outlier detection common tests

albertcthomas committed:
add tests for novelty ValueError

albertcthomas committed:
update plot_lof to plot_lof for outlier detection

albertcthomas committed:
[circle full] novelty detection example

albertcthomas commented:
BTW something's strange in the CI tests... the example in examples/covariance/plot_outlier_detection.py should fail as it is using the private method `_decision_function`, which I removed in the previous commits.

Running the example locally returns
````python
AttributeError: 'LocalOutlierFactor' object has no attribute '_decision_function'
````

albertcthomas commented:
I also suggest to remove this example (examples/covariance/plot_outlier_detection.py) since we now have examples/plot_anomaly_comparison.py. WDYT?

albertcthomas commented:
For plot_lof_outlier_detection.py, I refitted the model with novelty=True in order to show the level sets of the decision_function but that might confuse the users. We should maybe remove this and only illustrate the use of the negative_outlier_factor_ attribute.

agramfort commented:
ok but I would suggest to reuse the blob of text from the doc at the top.
At least some parts.
Check that you don't break any link in the doc.

albertcthomas commented:
> ok but I would suggest to reuse the blob of text from the doc at the top.
> At least some parts.
> Check that you don't break any link in the doc.

@agramfort you are saying this about removing plot_outlier_detection.py?

sklearn-lgtm commented:
This pull request **introduces 1 alert** when merging 25e93623f1ccf59d1902899d7234ea77cc1b2d3c into bfad4da2d06df36fcf4c2de173ec574107ab4457 - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-af360db77c447cbbb37d41fb63a18d2355e0daee)

**new alerts:**

* 1 for Variable defined multiple times

---

*Comment posted by [lgtm.com](https://lgtm.com)*

albertcthomas committed:
fix travis and outlier_detection example

albertcthomas committed:
add 2 components gaussian mixture with same variance in plot_anomaly_comparison to remove plot_outlier_detection

albertcthomas committed:
[doc build] rm plot_outlier_detection and update outlier detection doc

albertcthomas commented:
I'm almost done here, just need to fix a couple typos in the doc but Circleci returns the following error. I don't know if it's related to this PR.
```python
Unexpected failing examples:
/home/circleci/project/examples/gaussian_process/plot_gpr_co2.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/circleci/project/examples/gaussian_process/plot_gpr_co2.py", line 75, in <module>
    data = fetch_mldata('mauna-loa-atmospheric-co2').data
  File "/home/circleci/project/sklearn/datasets/mldata.py", line 154, in fetch_mldata
    mldata_url = urlopen(urlname)
  File "/home/circleci/miniconda/envs/testenv/lib/python3.6/urllib/request.py", line 223, in urlopen
    return opener.open(url, data, timeout)
  File "/home/circleci/miniconda/envs/testenv/lib/python3.6/urllib/request.py", line 532, in open
    response = meth(req, response)
  File "/home/circleci/miniconda/envs/testenv/lib/python3.6/urllib/request.py", line 642, in http_response
    'http', request, response, code, msg, hdrs)
  File "/home/circleci/miniconda/envs/testenv/lib/python3.6/urllib/request.py", line 570, in error
    return self._call_chain(*args)
  File "/home/circleci/miniconda/envs/testenv/lib/python3.6/urllib/request.py", line 504, in _call_chain
    result = func(*args)
  File "/home/circleci/miniconda/envs/testenv/lib/python3.6/urllib/request.py", line 650, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 500: Internal Server Error
```

agramfort commented:
push to restart the build. it's likely a random failure

albertcthomas commented:
Yes that's what I already tried. I will try again this evening. Thanks!

jnothman commented:
mldata has been having some down time lately...​

albertcthomas committed:
[doc build] sc in doc + add test for training scores

albertcthomas committed:
[doc build] sc and add table for LOF behavior in doc

albertcthomas commented:
This PR is ready for reviews. 1 question:
- Do we want to raise a warning if X_test is almost equal to X_train when using predict(X_test)/decision_function(X_test)/score_samples(X_test) in the novelty detection mode?

The doc is rendered at https://22700-843222-gh.circle-artifacts.com/0/doc/modules/outlier_detection.html

BTW using http://scikit-learn.org/circle?{BUILD_NUMBER} as indicated in the contributing section did not work...

jnothman commented:
yes, that url is broken since we started having circle runs on two python
versions. It is no longer mentioned in dev docs. use the browser script?

as to warning, I'm ambivalent. what kind of overhead does it add?

Review coming eventually... :/

albertcthomas commented:
Thanks @jnothman I should have checked the dev docs... where the new process is described very clearly.

For the warning, the training samples are already stored in the private attribute `_fit_X` but it might be (too) time consuming to use assert_array_almost_equal on big data sets...

agramfort commented on doc/modules/outlier_detection.rst:
```diff
@@ -44,19 +45,71 @@ inliers::
+might give useful results in these situations depending on the value of its
+hyperparameters.
+
+:class:`covariance.EllipticEnvelope` assumes the data is Gaussian and learns an ellipse. It thus degrades when the data is not unimodal. Notice however that this estimator is robust to outliers.
```
@albertcthomas line is too long.

agramfort commented on doc/modules/outlier_detection.rst:
```diff
@@ -236,20 +289,22 @@ where abnormal samples have different underlying densities.
 with respect to the surrounding neighborhood.
 
+When applying LOF for outlier detection, there are no ``predict``,
+``decision_function`` and ``score_samples`` methods but only a ``fit_predict`` method. The scores of abnormality of the training samples are accessible
```
line too long.

agramfort commented on doc/modules/outlier_detection.rst:
```diff
@@ -236,20 +289,22 @@ where abnormal samples have different underlying densities.
 
+When applying LOF for outlier detection, there are no ``predict``,
+``decision_function`` and ``score_samples`` methods but only a ``fit_predict`` method. The scores of abnormality of the training samples are accessible
+through the ``negative_outlier_factor_`` attribute.
```
Add a not that this is however possible when setting the novelty parameter to True. And put a cross-ref to the section that explains this.

agramfort commented on doc/modules/outlier_detection.rst:
```diff
@@ -236,20 +289,22 @@ where abnormal samples have different underlying densities.
 
-   * See :ref:`sphx_glr_auto_examples_neighbors_plot_lof.py` for
-     an illustration of the use of :class:`neighbors.LocalOutlierFactor`.
+   * See :ref:`sphx_glr_auto_examples_neighbors_plot_lof_outlier_detection.py` for an illustration of the use of :class:`neighbors.LocalOutlierFactor`.
```
too long.

agramfort commented on doc/modules/outlier_detection.rst:
```diff
@@ -236,20 +289,22 @@ where abnormal samples have different underlying densities.
 
-   * See :ref:`sphx_glr_auto_examples_covariance_plot_outlier_detection.py` for a
-     comparison with other anomaly detection methods.
+   * See :ref:`sphx_glr_auto_examples_plot_anomaly_comparison.py` for a comparison with other anomaly detection methods.
```
too long.

agramfort commented on doc/modules/outlier_detection.rst:
```diff
@@ -258,72 +313,42 @@ This strategy is illustrated below.
 
-.. topic:: Examples:
+To use :class:`neighbors.LocalOutlierFactor` for novelty detection, i.e.
+predict labels or compute the score of abnormality of new unseen data, you need to instantiate the estimator with the ``novelty`` parameter
```
line too long.

agramfort commented on examples/neighbors/plot_lof_novelty_detection.py:
```diff
@@ -0,0 +1,83 @@
+
+The Local Outlier Factor (LOF) algorithm is an unsupervised anomaly detection
+method which computes the local density deviation of a given data point with
+respect to its neighbors. It considers as outlier samples that have a
```
It considers as outliers the samples

agramfort commented on examples/neighbors/plot_lof_novelty_detection.py:
```diff
@@ -0,0 +1,83 @@
+detection.
+
+The number of neighbors considered, (parameter n_neighbors) is typically
+chosen 1) greater than the minimum number of objects a cluster has to contain,
```
chosen -> set to

agramfort replied:
objects -> samples.

also below.

agramfort commented on examples/neighbors/plot_lof_outlier_detection.py:
```diff
@@ -0,0 +1,85 @@
+
+The Local Outlier Factor (LOF) algorithm is an unsupervised anomaly detection
+method which computes the local density deviation of a given data point with
+respect to its neighbors. It considers as outlier samples that have a
```
same remark here as it's a copy paste.

agramfort commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1471,24 +1516,30 @@ def check_estimators_unfitted(name, estimator_orig):
     X, y = _boston_subset()
 
-    est = clone(estimator_orig)
+    estimator = clone(estimator_orig)
```
we tend to avoid these cosmetic changes...

albertcthomas replied:
ok... I thought it was ok when neighbouring code was modified...

albertcthomas committed:
address review

albertcthomas commented:
Thanks @agramfort! Sorry about all the lines that were too long. Is it written somewhere in the doc that lines in the documentation should be shorter than a given length?

Also I'm now wondering if it's a good idea to use LOF in the novelty detection mode to plot the decision boundary in plot_lof_outlier_detection example. This might confuse users. What did you think when looking at this example?

albertcthomas commented:
@jnothman should I remove LocalOutlierFactor as an example of transductive methods in the glossary?

jnothman commented:
I still think it's relevant there, since you need a switch to make it not.
perhaps extend the reference there

albertcthomas commented:
Actually I agree and that’s why I asked before making the change. It is still transductive by default so let’s keep the glossary unchanged.

jnothman commented on doc/modules/outlier_detection.rst:
```diff
@@ -258,72 +322,45 @@ This strategy is illustrated below.
+
+  When ``novelty`` is set to ``True`` be aware that you must only use
+  ``predict``, ``decision_function`` and ``score_samples`` on new unseen data
+  and not on the training samples as this would lead to wrong results.
```
Is there any sensible way to use it in a pipeline then?

albertcthomas replied:
Do you mean: what would be the correct way to use LOF in a Pipeline?

jnothman commented on doc/modules/outlier_detection.rst:
```diff
@@ -44,19 +45,73 @@ inliers::
+compares the score of abnormality of one sample with the scores of its
+neighbors.
+
+Finally, for the last data set, it is hard to say that one sample is more
```
We would usually reserve this level of detail for the example page, not the user guide

albertcthomas replied:
The previous version of this discussion was in the section "One-class SVM versus Elliptic Envelope versus Isolation Forest versus LOF" of the user guide to describe the results of the plot_outlier_detection example. But I can move this discussion to the plot_anomaly_comparison example and point the reader to the example for more details.

jnothman commented on examples/neighbors/plot_lof_outlier_detection.py:
```diff
@@ -0,0 +1,85 @@
+outlier detection and novelty detection and how to use LOF for novelty
+detection.
+
+The number of neighbors considered, (parameter n_neighbors) is typically
```
Remove comma

jnothman commented on sklearn/neighbors/lof.py:
```diff
@@ -250,26 +262,52 @@ def _predict(self, X=None):
         return is_inlier
 
-    def _decision_function(self, X):
+    def predict(self, X):
```
Should we use a property to make this disappear when not novelty?

albertcthomas replied:
I know the concept of properties but I'm not very familiar with it so it's not very clear to me how this would work. If you could clarify what you have in mind there is a good chance that I will be +1 to make this disappear when not novelty :).

albertcthomas replied:
Thanks. The goal of doing this is to have `hasattr` return False when not novelty. This is what you meant when you wrote "make this disappear when not novelty", i.e. make predict disappear when not novelty. Right?

albertcthomas replied:
In Python 3, `hasattr` returns False if an AttributeError is raised. This means that I need to raise an AttributeError if not novelty whereas we were raising a ValueError. That being said I'm also fine with AttributeError, as the existence of the attribute depends on the value of novelty.

jnothman replied:
But I think an AttributeError is raised if `__get__` fails with whichever error. But I might not be recalling correctly. 

Yes, do it for the other methods too

jnothman commented:
My question about pipeline was, I think, a bit misguided... but it would be
good to make the example or user guide more illustrative of novelty
detection applications. E.g. is it used for active learning or otherwise to
signal that you need to retrain a model?​

albertcthomas commented:
I don't know about active learning but novelty detection might indeed be used to signal that you need to retrain a model; for instance because the distribution of the data on which you want to predict is not the same as the one on which you trained your model. I think it is important to note that in such a case, the goal is to test whether P_train = P_test, whereas usually when you do novelty detection you fix a threshold (e.g. 5%) and you only estimate the corresponding density level set of P_train. That being said, it could be easier to compare one or several density level sets than two whole distributions. There was a [paper](https://papers.nips.cc/paper/4553-learning-high-density-regions-for-a-generalized-kolmogorov-smirnov-test-in-high-dimensional-data) at NIPS 2012 about a generalized KS test using One-Class SVMs. I could add one sentence or two and this reference to the documentation.

Another example of novelty detection application is the monitoring of events that should be investigated in priority because outside of a high density level set of the training data set.

albertcthomas committed:
remove comma in plot docstring

albertcthomas committed:
move details on comparison between outlier detection estimators from doc to example

albertcthomas committed:
sc in beginning of novelty and outlier detection doc + add distinction between outlier and novelty detection for anomaly detection

albertcthomas commented:
I think I addressed most of your comments @jnothman. For the novelty detection applications, most of the references I found are about anomaly detection. I thus clarified the introduction of the documentation, mentioned anomaly detection and explained the distinction between outlier and novelty detection for this application. Maybe @agramfort would also like to have a look.

albertcthomas committed:
sc in doc

albertcthomas committed:
use properties to make method disappear depending on novelty value

albertcthomas commented:
I'm not completely done yet.

sklearn-lgtm commented:
This pull request **fixes 2 alerts** when merging 3b0b5edc5e166c3176a539cd036ba1215c0dba96 into 1ff836438775cca5c480bde82b323651046c3fb2 - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-e91fbd1fb70e01f7f8b6c26a66840350a0a1c6a0)

**fixed alerts:**

* 2 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

albertcthomas committed:
remove novelty=True statements in estimator_checks and use check_estimator in test_lof

albertcthomas commented on sklearn/utils/estimator_checks.py:
```diff
@@ -446,6 +445,10 @@ def check_estimator_sparse_data(name, estimator_orig):
         try:
             with ignore_warnings(category=(DeprecationWarning, FutureWarning)):
                 estimator.fit(X, y)
+            if hasattr(estimator, 'novelty'):
```
I removed all the the `if hasattr(estimator, 'novelty')` from estimator_checks and I added a test in test_lof.py for `check_estimator(LocalOutlierFactor(novelty=True))`

albertcthomas commented:
I had to change ValueError to AttributeError for hasattr to return False.

albertcthomas commented:
I rebased on master to fix the conflicts.

albertcthomas commented:
Doc is rendered [here](https://27328-843222-gh.circle-artifacts.com/0/doc/modules/outlier_detection.html).

jnothman reviewed (approved):
To avoid regressions, please add tests that the  prediction methods are present iff novelty=True. Otherwise LGTM.

jnothman reviewed (commented):
Please add or update an entry in what's new.

jnothman commented:
Btw, great work on describing *novelty* vs *outlier*.

albertcthomas committed:
add non regression tests checking availability of prediction methods depending on the novelty parameter

albertcthomas committed:
add entry in whatsnew

albertcthomas commented:
Thanks @jnothman!

agramfort committed:
pass on docstrings

albertcthomas committed:
remove decision_function in outlier examplea and use negative_outlier_factor_

albertcthomas committed:
update whatsnew with author

albertcthomas commented:
Updated doc is rendered [here](https://27968-843222-gh.circle-artifacts.com/0/doc/modules/outlier_detection.html).

agramfort commented:
I am a fan of the new figure:

https://27968-843222-gh.circle-artifacts.com/0/doc/_images/sphx_glr_plot_lof_outlier_detection_0011.png

a nice copy of the wikipedia example

https://en.wikipedia.org/wiki/Local_outlier_factor#/media/File:LOF.svg

agramfort commented:
@albertcthomas can you rebase on master to restart appveyor

for a weird reason appveyor points to https://ci.appveyor.com/project/raghavrv/scikit-learn/build/1.0.10707 which is not the org appveyor account

albertcthomas commented:
appveyor still points to https://ci.appveyor.com/project/raghavrv/scikit-learn/build/1.0.10808 ...

agramfort commented:
Try opening a new PR

albertcthomas commented:
New PR #11569, using the correct appveyor account. Thanks!

albertcthomas commented:
I rebased on master to solve the conflicts and re-launch the tests.

GaelVaroquaux commented:
Actually, I terribly want to merge this guy, but I worry about the code coverage, and codecov is not running, as it is still waiting for appveyor (which will take forever).

albertcthomas commented:
I don’t understand why appveyor is failing. Same problem in PR #11555.

albertcthomas commented:
Just saw on gitter that people are aware of this issue :).

agramfort commented:
@albertcthomas you have some legit travis failures here.

albertcthomas commented:
I dont think they are coming from this PR, I will push a commit to restard CIs.

albertcthomas commented:
OK some of them are indeed coming from this PR, because of DeprecationWarning. I will try to fix them later in the day.

albertcthomas commented:
Doing what @glemaitre explained [here](https://github.com/scikit-learn/scikit-learn/pull/11553#issuecomment-405751708) should make travis happy.

albertcthomas commented:
Actually, @glemaitre what's the difference between
`@pytest.mark.filterwarnings` and `@ignore_warnings(category=DeprecationWarning)`?

albertcthomas committed:
ignore deprecationwarning for new lof tests

albertcthomas commented:
CIs are green but I don’t know why we don’t have the codecov checks.

agramfort commented:
good to go from my end.

Anyone for MRG+2?

GaelVaroquaux commented:
I would have liked codecov to run. I don't understand why it's not running. I'd like to check the coverage of this PR.

albertcthomas commented:
I restarted CIs to see if we can get code coverage.

albertcthomas commented:
@GaelVaroquaux codecov is green

glemaitre commented:
> Actually, @glemaitre what's the difference between
@pytest.mark.filterwarnings and @ignore_warnings(category=DeprecationWarning)?

We can make more fine-grain ignoring.

glemaitre commented:
It looks good to me. I would however change the deprecation warning to future warning for the outier methods.

albertcthomas commented:
Thanks @jnothman @agramfort @GaelVaroquaux and @glemaitre for the help and the reviews!
